File: trivia_bot.js
// ============================================================
// RPOW2 TRIVIA BOT - Browser Console Version
// ============================================================
//
// Cara pakai:
//   1. Buka https://trivia.rpow2.com/ di browser (sudah login)
//   2. F12 → Console
//   3. Ganti myEmail & GROQ_KEY di bawah
//   4. Paste semua → Enter
//   5. Stop: ketik `stopBot()` di console
//
// Kenapa via browser console (bukan Python):
//   Trivia API pakai Cloudflare yang block IP datacenter.
//   Di browser pakai `credentials: "include"` → otomatis pakai
//   cookie browser → Cloudflare ga block.
//
// Update KNOWN saat bot salah:
//   KNOWN["kata_kunci_soal"] = INDEX_BENAR;
// ============================================================

const api       = "https://api.rpow2.com";
const myEmail   = "youremail@gmail.com";                                // ← GANTI email kamu
const GROQ_KEY  = "gsk_YOUR_GROQ_API_KEY_HERE";                         // ← dari console.groq.com (gratis)
const MIN_BET   = 0;  // set ke 100000000 (0.1 RPOW) untuk filter bet gede

let running = true, wins = 0, losses = 0, timeouts = 0;
function stopBot() { running = false; console.log(`STOPPED W=${wins} L=${losses} ⏳=${timeouts}`); }

// ============================================================
// KNOWN ANSWERS - soal yang pernah salah, supaya tidak salah lagi
// ============================================================
const KNOWN = {
  // RPOW / Bitcoin / Crypto
  "hal finney":           0,
  "rpow":                 0,
  "satoshi":              0,
  "bitcoin":              0,
  "sha-256":              1,
  "proof of work":        0,

  // Sudah pernah salah - sekarang dihardcode
  "whistler":             0,   // Windows XP
  "windows xp":           0,
  "kuwait":               1,   // 9 islands
  "how many islands":     1,
  "not exclusively female": 0, // Clefairy
  "clefairy":             0,
  "battery coined":       2,   // Benjamin Franklin
  "benjamin franklin":    2,
  "electrical storage":   2,
  "sonic 2 level select": 2,   // Lead Programmer's birthday
  "intel hd":             1,   // HD Graphics 500 (after Broadwell)
  "intel hd broadwell":   1,
  "zeptometre":           0,   // 1,000,000
  "femtometre":           0,
  "rebadged":             3,   // Isuzu Trooper
  "badge-engineered":     3,
  "warframe":             3,   // Mawfish
  "mawfish":              3,
  "cetus":                3,
  "eats unwary natives":  3,

  // Umum
  "sega mascot":          2,   // Sonic
  "sonic the hedgehog":   2,
  "linus torvalds":       0,   // Linux
  "linux":                0,
  "van gogh":             3,
  "cut off his ear":      3,
  "coulrophobia":         3,   // Clowns
  "spanish flu":          1,   // 1-3%
  "ridley scott":         2,
  "alien":                2,
  "evangelion":           0,   // A Cruel Angel's Thesis
};

// ============================================================
// HTTP helpers (pakai credentials:"include" biar Cloudflare ga block)
// ============================================================
async function req(method, path, body) {
  const res = await fetch(api + path, {
    method,
    credentials: "include",
    headers: {
      "content-type": "application/json",
      "accept":       "application/json"
    },
    body: body ? JSON.stringify(body) : undefined
  });
  return res.json();
}

// ============================================================
// Answer engine
//   1. KNOWN dict (instant)
//   2. Groq AI (llama-3.3-70b, fast & decent)
//   3. Fallback [0]
// ============================================================
async function askAI(question, choices) {
  // 1. KNOWN dict check
  const ql = question.toLowerCase();
  for (const [kw, idx] of Object.entries(KNOWN)) {
    if (ql.includes(kw) && idx < choices.length) {
      console.log(`  known[${idx}] ${choices[idx]}`);
      return idx;
    }
  }

  // 2. Groq AI
  try {
    const res = await fetch("https://api.groq.com/openai/v1/chat/completions", {
      method: "POST",
      headers: {
        "content-type":  "application/json",
        "authorization": `Bearer ${GROQ_KEY}`
      },
      body: JSON.stringify({
        model: "llama-3.3-70b-versatile",
        messages: [{
          role: "user",
          content: `Trivia question. Reply ONLY with 0, 1, 2, or 3. No explanation.

Q: ${question}
[0] ${choices[0]}
[1] ${choices[1]}
[2] ${choices[2]}
[3] ${choices[3]}`
        }],
        max_tokens: 3,
        temperature: 0
      })
    });
    const d = await res.json();
    const idx = parseInt(d.choices?.[0]?.message?.content?.trim());
    if (idx >= 0 && idx <= 3) {
      console.log(`  AI[${idx}] ${choices[idx]}`);
      return idx;
    }
  } catch (e) {
    console.log("  AI err:", e.message);
  }

  console.log("  fallback[0]");
  return 0;
}

async function sleep(ms) { return new Promise(r => setTimeout(r, ms)); }

// ============================================================
// Main match loop
// ============================================================
async function playOne() {
  const lobby = await req("GET", "/api/trivia/lobby");

  // Filter: bet >= MIN_BET, bukan akun sendiri
  let players = (lobby.players || [])
    .filter(p => p.account_email !== myEmail && p.bet_base_units >= MIN_BET)
    .sort((a, b) => b.bet_base_units - a.bet_base_units);

  // Fallback: kalau tidak ada yang memenuhi MIN_BET, ambil bet terbesar yang ada
  if (!players.length) {
    const all = (lobby.players || [])
      .filter(p => p.account_email !== myEmail)
      .sort((a, b) => b.bet_base_units - a.bet_base_units);
    if (!all.length) { console.log("lobby empty..."); await sleep(5000); return; }
    players = [all[0]];
  }

  const t   = players[0];
  const bet = t.bet_base_units / 1e9;
  console.log(`\nvs @${t.x_handle} bet=${bet} W/L=${t.matches_won}/${t.matches_lost}`);

  // Start match
  const match = await req("POST", "/api/trivia/matches/start", { session_id: t.session_id });
  if (!match.match_id && !match.id) {
    console.log("start fail:", match);
    await sleep(3000);
    return;
  }

  const mid = match.match_id || match.id;
  const q   = match.question || "";
  const ch  = match.choices  || [];
  console.log(`Q: ${q}`);
  ch.forEach((c, i) => console.log(`  [${i}] ${c}`));

  // Pick & submit answer
  const idx = await askAI(q, ch);
  console.log(`>>> [${idx}] ${ch[idx]}`);
  await req("POST", `/api/trivia/matches/${mid}/answer`, { choice_idx: idx });

  // Poll hasil (500ms × 40 = 20 detik max)
  let winner = null, ci = null;
  for (let i = 0; i < 40; i++) {
    await sleep(500);
    const st = await req("GET", `/api/trivia/matches/${mid}`);
    winner = st.winner_email;
    ci     = st.correct_choice_idx;
    if (winner) break;
  }

  if (ci !== null && ci !== undefined) console.log(`correct:[${ci}] ${ch[ci]}`);

  if (winner === myEmail) {
    wins++;
    console.log(`✅ WON +${bet * 2} RPOW (W=${wins} L=${losses})`);
  } else if (winner) {
    losses++;
    console.log(`❌ LOST (W=${wins} L=${losses})`);
    // Auto-learn: simpan keyword dari soal yang salah → jawaban benar
    if (ci !== null && ci !== undefined) {
      const words = (q.toLowerCase().match(/[a-z]{5,}/g) || []).slice(0, 2);
      words.forEach(w => {
        KNOWN[w] = ci;
        console.log(`  learned:"${w}"→[${ci}]`);
      });
    }
  } else {
    timeouts++;
    console.log(`⏳ timeout`);
  }

  const me = await req("GET", "/me");
  console.log(`balance: ${(me.balance_base_units / 1e9).toFixed(3)} RPOW`);
}

// ============================================================
// Main loop
// ============================================================
(async () => {
  const me = await req("GET", "/me");
  console.log(`=== RPOW2 TRIVIA BOT ===`);
  console.log(`${me.email}  balance: ${(me.balance_base_units / 1e9).toFixed(3)} RPOW`);
  console.log(`KNOWN: ${Object.keys(KNOWN).length} entries | MIN_BET: ${MIN_BET / 1e9} RPOW`);
  console.log(`stop: stopBot()`);

  while (running) {
    try {
      await playOne();
    } catch (e) {
      console.error("err:", e.message);
      await sleep(2000);
    }
  }
})();

File: trivia_bot.py
#!/usr/bin/env python3
"""
trivia_bot.py - Auto-join & answer rpow2 Trivia matches.

Flow:
  1. GET /api/trivia/lobby   → lihat open sessions
  2. Pilih session bet_base_units terkecil
  3. POST /api/trivia/matches/start  { session_id }
  4. Baca question + choices dari response
  5. Jawab dengan choice_idx yang benar
  6. Ulangi

Usage (di PC/HP yang IP-nya tidak diblock Cloudflare):
  export RPOW2_COOKIE='cf_clearance=...; rpow_session=...'
  python3 trivia_bot.py              # loop terus
  python3 trivia_bot.py --once       # 1 match lalu exit
  python3 trivia_bot.py --dry-run    # lihat lobby tanpa join
"""

from __future__ import annotations

import argparse
import json
import logging
import os
import re
import sys
import time
import urllib.error
import urllib.request
from typing import Any

API          = "https://api.rpow2.com"
TRIVIA_ORIGIN = "https://trivia.rpow2.com"
UA = (
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
    "AppleWebKit/537.36 (KHTML, like Gecko) "
    "Chrome/124.0.0.0 Safari/537.36"
)

log = logging.getLogger("trivia")
BASE_UNITS = 1_000_000_000


def fmt(bu: int | str) -> str:
    try:
        v = int(bu)
    except Exception:
        return str(bu)
    whole, frac = divmod(v, BASE_UNITS)
    return f"{whole}.{frac:09d} RPOW"


# ---------- HTTP ---------------------------------------------------------
class Client:
    def __init__(self, cookie: str) -> None:
        if not cookie or "=" not in cookie:
            raise SystemExit(
                "RPOW2_COOKIE not set.\n"
                "Buka trivia.rpow2.com di browser → F12 → Network\n"
                "→ klik request ke api.rpow2.com → copy Cookie header\n"
                "→ export RPOW2_COOKIE='cf_clearance=...; rpow_session=...'"
            )
        self.cookie = cookie
        self._last  = 0.0

    def _gap(self, min_gap: float = 0.3) -> None:
        dt = time.monotonic() - self._last
        if dt < min_gap:
            time.sleep(min_gap - dt)
        self._last = time.monotonic()

    def req(self, method: str, path: str,
            body: dict | None = None, timeout: float = 20) -> tuple[int, Any]:
        self._gap()
        headers = {
            "cookie":         self.cookie,
            "accept":         "application/json",
            "user-agent":     UA,
            "origin":         TRIVIA_ORIGIN,
            "referer":        f"{TRIVIA_ORIGIN}/",
            "sec-fetch-dest": "empty",
            "sec-fetch-mode": "cors",
            "sec-fetch-site": "same-site",
        }
        data = None
        if body is not None:
            data = json.dumps(body).encode()
            headers["content-type"] = "application/json"
        r = urllib.request.Request(
            API + path, data=data, headers=headers, method=method)
        try:
            with urllib.request.urlopen(r, timeout=timeout) as resp:
                raw = resp.read().decode() or "{}"
                return resp.status, json.loads(raw)
        except urllib.error.HTTPError as e:
            raw = e.read().decode()
            try:    return e.code, json.loads(raw)
            except: return e.code, {"raw": raw[:300]}

    def lobby(self) -> list:
        s, r = self.req("GET", "/api/trivia/lobby")
        if s != 200:
            raise RuntimeError(f"lobby {s}: {r}")
        return r if isinstance(r, list) else r.get("players", [])

    def stats(self) -> dict:
        s, r = self.req("GET", "/api/trivia/stats")
        return r if s == 200 else {}

    def me(self) -> dict:
        s, r = self.req("GET", "/me")
        if s != 200:
            raise RuntimeError(f"me {s}: {r}")
        return r

    def start_match(self, session_id: str) -> dict:
        s, r = self.req("POST", "/api/trivia/matches/start",
                        {"session_id": session_id})
        if s != 200:
            raise RuntimeError(f"start_match {s}: {r}")
        return r

    def answer(self, match_id: str, choice_idx: int) -> dict:
        s, r = self.req("POST", f"/api/trivia/matches/{match_id}/answer",
                        {"choice_idx": choice_idx})
        if s != 200:
            raise RuntimeError(f"answer {s}: {r}")
        return r

    def match_state(self, match_id: str) -> dict:
        s, r = self.req("GET", f"/api/trivia/matches/{match_id}")
        return r if s == 200 else {}


# ---------- Answer engine ------------------------------------------------
# Known answers: key = lowercase question, value = correct choice_idx.
# Built from rpow2 trivia topics (crypto, Hal Finney, RPOW history).

KNOWN: dict[str, int] = {
    # RPOW / rpow2 specific
    "who created the original rpow system in 2004?": 0,          # Hal Finney
    "who created the original rpow (reusable proof of work) system in 2004?": 0,
    "what hash function did hal finney's original rpow use?": 1,  # SHA-1
    "rpow2 is a tribute to which bitcoin pioneer?": 0,            # Hal Finney
    "what does rpow stand for?": 2,                               # Reusable Proof of Work
    "what is the total supply cap of rpow2?": 1,                  # 21 million
    "what blockchain is srpow bridged to?": 0,                    # Solana
    "what proof-of-work algorithm does rpow2 use?": 1,            # SHA-256
    "how many base units equal 1 rpow?": 0,                       # 1,000,000,000
    # Bitcoin
    "who invented bitcoin?": 0,                                   # Satoshi Nakamoto
    "what year was the bitcoin whitepaper published?": 1,         # 2008
    "what year was bitcoin created?": 1,                          # 2008/2009
    "what is the smallest unit of bitcoin called?": 2,            # satoshi
    "how many satoshis are in one bitcoin?": 3,                   # 100,000,000
    "what consensus mechanism does bitcoin use?": 0,              # proof of work
    "what does sha stand for?": 2,                                # Secure Hash Algorithm
    "how often does the bitcoin halving occur?": 2,               # every ~4 years
    "what is the bitcoin block time target?": 1,                  # 10 minutes
    "what is the maximum supply of bitcoin?": 0,                  # 21 million
    "what language is bitcoin's script written in?": 1,           # Script
    # Crypto general
    "what is a nonce in proof of work?": 0,
    "what is a merkle tree used for in blockchain?": 0,
    "what is a 51% attack?": 1,
    "what is the ethereum virtual machine called?": 1,            # EVM
    "what does defi stand for?": 0,                               # Decentralized Finance
    "what is a smart contract?": 2,
}


def normalize(q: str) -> str:
    return re.sub(r"\s+", " ", q.strip().lower())


def pick(question: str, choices: list[str]) -> int:
    """Return best choice_idx using known answers → heuristics → fallback."""
    nq = normalize(question)

    # 1. Exact match
    if nq in KNOWN:
        idx = KNOWN[nq]
        log.info("known [%d] %s", idx, choices[idx] if idx < len(choices) else "?")
        return idx

    # 2. Partial match on known keys
    for k, idx in KNOWN.items():
        if len(k) > 20 and (k in nq or nq in k):
            log.info("partial-match [%d] %s", idx, choices[idx] if idx < len(choices) else "?")
            return idx

    # 3. Keyword heuristics: scan choices for strong signals
    cl = [c.lower() for c in choices]
    ql = nq

    heuristics = [
        # (keyword in question, term in choice, priority)
        ("hal finney",       ["hal finney", "finney"],          0),
        ("satoshi",          ["satoshi nakamoto", "nakamoto"],   0),
        ("sha-256",          ["sha-256", "sha256"],              0),
        ("sha-1",            ["sha-1", "sha1"],                  0),
        ("solana",           ["solana", "sol"],                  0),
        ("proof of work",    ["proof of work", "pow"],           0),
        ("21 million",       ["21 million", "21,000,000"],       0),
        ("1 billion",        ["1,000,000,000", "1 billion"],     0),
        ("reusable",         ["reusable proof", "rpow"],         0),
        ("bitcoin",          ["bitcoin", "btc"],                 0),
        ("ethereum",         ["ethereum", "eth"],                0),
        ("merkle",           ["merkle"],                         0),
        ("10 minutes",       ["10 minutes", "10 min"],           0),
        ("2008",             ["2008"],                           0),
        ("2009",             ["2009"],                           0),
        ("100,000,000",      ["100,000,000", "100 million"],     0),
    ]

    for keyword, terms, _ in heuristics:
        if keyword in ql:
            for term in terms:
                for i, c in enumerate(cl):
                    if term in c:
                        log.info("heuristic '%s' → [%d] %s", term, i, choices[i])
                        return i

    # 4. Fallback
    log.warning("unknown Q, guessing [0]. Add to KNOWN dict if wrong.")
    log.warning("Q: %s", question)
    for i, ch in enumerate(choices):
        log.warning("  [%d] %s", i, ch)
    return 0


# ---------- Main loop ----------------------------------------------------
def run(c: Client, once: bool, dry_run: bool) -> None:
    try:
        me = c.me()
        log.info("account: %s  balance: %s",
                 me.get("email"), fmt(me.get("balance_base_units", 0)))
    except Exception as e:
        log.warning("could not load account: %s", e)
        me = {}

    my_email = me.get("email", "")

    try:
        st = c.stats()
        if st:
            log.info("trivia global: matches=%s  volume=%s",
                     st.get("total_matches", "?"),
                     fmt(st.get("total_volume_base_units", 0)))
    except Exception:
        pass

    wins = losses = 0

    while True:
        try:
            # --- Lobby ---
            players = c.lobby()
            open_sessions = [
                p for p in players
                if p.get("account_email") != my_email
                and int(p.get("bankroll_remaining_base_units", 0)) > 0
            ]

            if not open_sessions:
                log.info("lobby empty, waiting 10s...")
                if dry_run or once:
                    return
                time.sleep(10)
                continue

            # Sort by bet ascending
            open_sessions.sort(key=lambda p: int(p.get("bet_base_units", 0)))
            log.info("=== LOBBY (%d open sessions) ===", len(open_sessions))
            for p in open_sessions[:8]:
                log.info("  %-30s  bet=%-25s  bankroll=%-25s  W/L=%s/%s",
                         (p.get("x_handle") or p.get("account_email", "?"))[:30],
                         fmt(p.get("bet_base_units", 0)),
                         fmt(p.get("bankroll_remaining_base_units", 0)),
                         p.get("matches_won", 0),
                         p.get("matches_lost", 0))

            target     = open_sessions[0]
            session_id = (target.get("session_id")
                          or target.get("id")
                          or target.get("account_email"))
            bet        = fmt(target.get("bet_base_units", 0))
            opponent   = (target.get("x_handle")
                          or target.get("account_email", "?"))

            if dry_run:
                log.info("[dry-run] would challenge: %s  bet=%s", opponent, bet)
                return

            log.info("→ challenging %s  bet=%s  session_id=%s",
                     opponent, bet, session_id)

            # --- Start match ---
            match      = c.start_match(session_id)
            match_id   = match.get("match_id") or match.get("id")
            question   = match.get("question", "")
            choices    = match.get("choices", [])
            deadline   = match.get("deadline_at", "")
            log.info("match %s | deadline: %s", match_id, deadline)
            log.info("Q: %s", question)
            for i, ch in enumerate(choices):
                log.info("  [%d] %s", i, ch)

            # --- Pick & submit answer ---
            idx    = pick(question, choices)
            result = c.answer(match_id, idx)
            log.info("answered [%d] → %s", idx, result)

            # --- Wait for result (poll up to 20s) ---
            winner = result.get("winner_email")
            for _ in range(20):
                if winner:
                    break
                time.sleep(1)
                state  = c.match_state(match_id)
                winner = state.get("winner_email")

            if winner == my_email:
                wins  += 1
                prize  = int(target.get("bet_base_units", 0)) * 2
                log.info("✅ WON +%s  (W=%d L=%d)", fmt(prize), wins, losses)
            elif winner:
                losses += 1
                log.info("❌ LOST  (W=%d L=%d)", wins, losses)
                # Learn from loss: add correct answer to KNOWN
                correct = result.get("correct_choice_idx")
                if correct is not None and question:
                    nq = normalize(question)
                    KNOWN[nq] = int(correct)
                    log.info("learned: Q='%s' → [%d]", question[:60], correct)
            else:
                log.info("⏳ no result (W=%d L=%d)", wins, losses)

            if once:
                return

            time.sleep(3)

        except KeyboardInterrupt:
            log.info("stopped. W=%d L=%d", wins, losses)
            return
        except RuntimeError as e:
            log.error("error: %s — retrying in 5s", e)
            time.sleep(5)
        except Exception as e:
            log.exception("unexpected: %s", e)
            time.sleep(5)


def main() -> None:
    logging.basicConfig(
        level=logging.INFO,
        format="%(asctime)s %(levelname)s %(message)s",
        datefmt="%H:%M:%S",
    )
    ap = argparse.ArgumentParser(description="rpow2 trivia bot")
    ap.add_argument("--once",    action="store_true",
                    help="play 1 match then exit")
    ap.add_argument("--dry-run", action="store_true",
                    help="show lobby only, don't join any match")
    args = ap.parse_args()

    cookie = os.environ.get("RPOW2_COOKIE", "").strip()
    c = Client(cookie)
    run(c, once=args.once, dry_run=args.dry_run)


if __name__ == "__main__":
    main()
